### PR TITLE
Perm enum

### LINKED ISSUE #2986: Use enum as permission code
### What I'm trying to achieve
Code transparency and type safety.

### Describe a proposed solution
Return `enum` instead of `string` as permission code.

### PR DESCRIPTION
**Pull request by the-bionic**
I want to merge this change because...
resolves #2986

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] GraphQL schema and type definitions are up to date.

### CODE REVIEW

File: tests/api/test_shop.py [hunk 1]
```diff
@@ -12,7 +13,7 @@ def test_query_authorization_keys(
     query = """
     query {
         shop {
-            authorizationKeys {
+            s {
```
Comment (Pacu2) [inline tests/api/test_shop.py]: Not sure why this change is needed?

**Comment by Pacu2** [comment]
Could you please fix codeclimate issues?

**Comment by the-bionic** [comment]
Yes, I am fixing that.

**Comment by Pacu2** [comment]
The ReactJS part needs adjustment, are you able to handle that or shall we?

**Comment by the-bionic** [comment]
Alright. I'll handle it.

**Comment by the-bionic** [comment]
It just occurred to me that this is going to need some reworking as well:

``` python
def format_permissions_for_display(permissions):
    """Transform permissions queryset into PermissionDisplay list.
    Keyword arguments:
    permissions - queryset with permissions
    """
    return [
        PermissionDisplay(
            code='.'.join(
                [permission.content_type.app_label, permission.codename]), name=permission.name) 
                for permission in permissions]
```
to return an enum so the frontend can perform its permission checks. Right now it is comparing ```string``` to ```PermissionEnum```, which is crashing it.

Here is what I came up with:
``` python
formatted_permissions = []
    for permission in permissions:
        codename = '.'.join(
            [permission.content_type.app_label, permission.codename])
        formatted_permissions.append(
            PermissionDisplay(
                code=graphene.Enum(
                    'PermissionEnum', str_to_enum(codename), codename),
                name=permission.name))
    return formatted_permissions
```

File: saleor/graphql/core/types/common.py [hunk 2]
```diff
@@ -59,8 +61,13 @@ class LanguageDisplay(graphene.ObjectType):
     language = graphene.String(description='Language.', required=True)
 
 
+PermissionEnum = graphene.Enum(
+    'PermissionEnum', [(str_to_enum(codename), codename)
```
Comment (Pacu2) [inline saleor/graphql/core/types/common.py]: I think we should split it by the dot here, as we did in `split_permission_codename` function, so the final permission will be
`MANAGE_USERS` 
instead of 
`ACCOUNT_MANAGE_USERS`

Comment (the-bionic) [reply saleor/graphql/core/types/common.py]: I'll resolve that.

**Comment by the-bionic** [comment]
``` python
def format_permissions_for_display(permissions):
    """Transform permissions queryset into PermissionDisplay list.
    Keyword arguments:
    permissions - queryset with permissions
    """
    return [
        PermissionDisplay(
            code='.'.join(
                [permission.content_type.app_label, permission.codename]),
name=permission.name) for permission in permissions]
```
This is going to need some modification. Because the frontend is going to be making its checks based on the ```PermissionEnum```, and not on ```strings```.

I was doing this:
``` python
formatted_permissions = []
    for permission in permissions:
        codename = '.'.join(
            [permission.content_type.app_label, permission.codename])
        formatted_permissions.append(
            PermissionDisplay(
                code=graphene.Enum(
                    'PermissionEnum', str_to_enum(codename), codename),
                name=permission.name))
    return formatted_permissions
```
What do you think?

**Comment by Pacu2** [comment]
@the-bionic If the result is as expected, I'm fine with that

**Comment by Pacu2** [comment]
Or actually, you can just use `getattr` and existing `PermissionEnum` 
Like `code=getattr(PermissionEnum, codename)`

**Comment by the-bionic** [comment]
I used ```Permission.get(codename)``` which does about the same thing.

**Comment by the-bionic** [comment]
Thanks! I am on it.

I have reverted the snapshot I updated to what it was before. I will investigate further and let you know you know if I need help.

**Comment by the-bionic** [comment]
```npm run storybook``` runs without errors, but 2 tests are still failing though.

**Comment by Pacu2** [comment]
@the-bionic In such case, please take a look if stories in storybook are rendered properly, especially configuration page stories.

**Comment by the-bionic** [comment]
@Pacu2 it renders okay (I am not sure how it would look when it doesn't.) After inspecting the differences in the snapshot, it turns out
```
<input
                              class="MuiSwitchBase-input-id"
    -                         name="account.impersonate_users"
    +                         name="IMPERSONATE_USERS"
                              type="checkbox"
                            />
```

That is happening in an input tag. That is what is breaking the test.

**Comment by Pacu2** [comment]
@the-bionic I think that's a correct behavior, as we've changed strings to Enums, we should adjust those tests and are good to go.

**Comment by the-bionic** [comment]
It can be inspected here, too:
https://travis-ci.org/mirumee/saleor/jobs/440580916

**Comment by the-bionic** [comment]
Great. I'll update the snapshot. 

You can also see here for the full details: https://travis-ci.org/mirumee/saleor/jobs/440580916

**Comment by Pacu2** [comment]
@dominik-zeglen we good?
